### PR TITLE
More efficient subscriptions

### LINKED ISSUE #2547: Efficient Subscriptions
[comment]: # (Please replace ... with your information. Remove < and >)

### Description

Subscriptions in our should be as effective as possible, forming a tree with potentially many simple, fast (lookup only) Level-1 `re-frame` subscriptions which are recalculated every-time the `app-db` is swapped, but less more complicated, slower subscriptions, which take those Level-1 subscriptions as input and so are recalculated only when necessary, see - https://github.com/status-im/ideas/issues/31

### PR DESCRIPTION
**Pull request by janherich**
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

Fixes #2547 

### Summary:

* Duplicate subscriptions removed (there were multiple instances of subscribing to particular chat)
* Ineffective Level-1 subscriptions rewritten to leverage efficient subscription graph, particularly in `status-im.chat.subs` and `status-im.ui.screen.discover.subs` namespaces, where it was worst
* Some smaller naming changes

I refrained from making other changes then to subscriptions to make the PR minimal, even so, it's not exactly super small footprint. 
There are many, many opportunities to significantly reduce code size and increase consistency/efficiency especially in the chat module, but I will wait till @alwx 's changes. 

### Steps to test:
- Open Status
- Throughly test the discover functionality
- Visit every screen of the app to catch potential errors

status: ready

### CODE REVIEW

File: src/status_im/chat/events/input.cljs [hunk 1]
```diff
@@ -64,6 +63,9 @@
       (and dapp? (str/blank? chat-text))
       (assoc-in [:chats current-chat-id :parameter-boxes :message] nil))))
 
+;; TODO janherich: this is super fragile and won't work at all for group chats with bots.
+;; The proper way how to do it is to check each chat participant and call `:on-message-input-change`
+;; jail function in each participant's jail
```
Comment (jeluard) [inline src/status_im/chat/events/input.cljs:68]: Can you create an issue to track this?

File: src/status_im/chat/models/input.cljs [hunk 2]
```diff
@@ -96,45 +96,44 @@
...
-         (when-let [{{:keys [message-id]} :request :as command} (get available-commands-responses (subs command-name 1))]
-           {:command  command
-            :metadata (if (and (nil? (:to-message-id input-metadata)) message-id)
-                        (assoc input-metadata :to-message-id message-id)
-                        input-metadata)
-            :args     (->> (if (empty? seq-arguments)
+  ([{:keys [input-text input-metadata seq-arguments] :as current-chat} commands responses]
+   (let [[command-name :as command-args] (split-command-args input-text)]
+     (when-let [{{:keys [message-id]} :request :as command} (and command-name
+                                                                 (starts-as-command? command-name)
```
Comment (jeluard) [inline src/status_im/chat/models/input.cljs:110]: Extracting some functions would improve readability

File: src/status_im/chat/models/input.cljs [hunk 3]
```diff
@@ -169,7 +168,7 @@
   (let [input-text    (get-in db [:chats current-chat-id :input-text])
         seq-arguments (get-in db [:chats current-chat-id :seq-arguments])
         selection     (get-in db [:chat-ui-props current-chat-id :selection])
-        chat-command  (selected-chat-command db current-chat-id)]
+        chat-command  (selected-chat-command db)]
     (current-chat-argument-position chat-command input-text selection seq-arguments)))
 
 (defn command-completion
```
Comment (jeluard) [inline src/status_im/chat/models/input.cljs:174]: Do we have tests for this?

File: src/status_im/chat/subs.cljs [hunk 4]
```diff
@@ -186,27 +171,30 @@
...
+  :<- [:get-current-chat]
+  (fn [current-chat]
+    (:all-loaded? current-chat)))
 
 (reg-sub :photo-path
   :<- [:get-contacts]
   (fn [contacts [_ id]]
     (:photo-path (contacts id))))
 
+;; TODO janherich: this is just bad and horribly ineffecient (always sorting to get last msg +
```
Comment (jeluard) [inline src/status_im/chat/subs.cljs]: Can you create an issue to track this?

File: src/status_im/ui/screens/contacts/subs.cljs [hunk 5]
```diff
@@ -65,7 +65,7 @@
     (filter-not-group-contacts group-contacts contacts)))
 
 (reg-sub
-  :all-added-group-contacts-with-limit
+    :all-added-group-contacts-with-limit
```
Comment (jeluard) [inline src/status_im/ui/screens/contacts/subs.cljs]: Alignement?

File: src/status_im/ui/screens/discover/subs.cljs [hunk 6]
```diff
@@ -2,18 +2,33 @@
...
+
+(reg-sub :get-discover-search-tags
+  (fn [db]
+    (:discover-search-tags db)))
+
+(reg-sub :get-tags
+  (fn [db]
+    (:tags db)))
+
+(defn- calculate-priority [chats current-public-key contacts
```
Comment (jeluard) [inline src/status_im/ui/screens/discover/subs.cljs]: Do we have tests for this?

File: src/status_im/ui/screens/discover/subs.cljs [hunk 7]
```diff
@@ -25,44 +40,53 @@
...
-          current-tag (:current-tag db)
-          tags        (:discover-search-tags db)]
-      (get-discoveries-by-tags discoveries current-tag tags))))
+  :<- [:get-discoveries]
+  :<- [:get-current-tag]
+  :<- [:get-discover-search-tags]
+  (fn [[discoveries current-tag discover-search-tags]]
+    (get-discoveries-by-tags discoveries current-tag discover-search-tags)))
 
+;; TODO janherich: is this really what we want, what about `state-of-dapps` group ?
```
Comment (jeluard) [inline src/status_im/ui/screens/discover/subs.cljs]: Can we clarify this in the chat, and create an issue if needed?

**Review by jeluard — CHANGES_REQUESTED** [review]

**Comment by jeluard** [comment]
@janherich Can you take this opportunity to start writing down a guideline for efficient subscriptions?

Also can you create issues associated to your TODOs comments when relevant?

File: src/status_im/bots/subs.cljs [hunk 8]
```diff
@@ -2,13 +2,11 @@
   (:require [re-frame.core :as re-frame]
             [status-im.chat.models.input :as input-model]))
 
-(re-frame/reg-sub
-  :bot-db
+(re-frame/reg-sub :bot-db
```
Comment (flexsurfer) [inline src/status_im/bots/subs.cljs]: in cursive this will be aligned bad, fn will be under :bot-db, let's use previous alignment

**Review by flexsurfer — CHANGES_REQUESTED** [review]
as i can see here is not only subscriptions changes?

File: src/status_im/ui/screens/discover/subs.cljs (hunk 7 above)
Comment (janherich) [reply src/status_im/ui/screens/discover/subs.cljs]: This is list of DApps on discover screen, nothing to do with chat

File: src/status_im/ui/screens/discover/subs.cljs (hunk 6 above)
Comment (janherich) [reply src/status_im/ui/screens/discover/subs.cljs]: No we don't, but it's a good opportunity to cover it with tests.

File: src/status_im/chat/models/input.cljs (hunk 3 above)
Comment (janherich) [reply src/status_im/chat/models/input.cljs:174]: Yes, we have tests for the `selected-chat-command` function

File: src/status_im/chat/events/input.cljs (hunk 1 above)
Comment (janherich) [reply src/status_im/chat/events/input.cljs:68]: Sure

File: src/status_im/chat/subs.cljs (hunk 4 above)
Comment (janherich) [reply src/status_im/chat/subs.cljs]: We already have general issue for that - https://app.zenhub.com/workspace/o/status-im/status-react/issues/1875

File: src/status_im/bots/subs.cljs (hunk 8 above)
Comment (janherich) [reply src/status_im/bots/subs.cljs]: I will revert it

File: src/status_im/ui/screens/discover/subs.cljs (hunk 7 above)
Comment (jeluard) [reply src/status_im/ui/screens/discover/subs.cljs]: I meant in slack :) To clarify if this is indeed what we really want.

**Comment by janherich** [comment]
@flexsurfer It's not subs-only changes only when necessary - for example when changing the function hooked on subscription which unnecessary takes the whole db as argument (such as `selected-chat-command`) I had to change the function signature to only take what's necessary + change the function call in other places (handlers) where it's used. 
But it's still driven by the need to have efficient subscriptions and there are only minimal changes necessary to make it work.

**Review by flexsurfer — APPROVED** [review]

**Review by jeluard — APPROVED** [review]

File: src/status_im/ui/screens/discover/subs.cljs (hunk 7 above)
Comment (goranjovic) [reply src/status_im/ui/screens/discover/subs.cljs]: Intentional. It is a temporary list of dapps, until idea `#14` is implemented.

**Review by goranjovic — APPROVED** [review]

File: src/status_im/ui/screens/discover/subs.cljs [hunk 9]
```diff
@@ -2,18 +2,25 @@
   (:require [re-frame.core :refer [reg-sub]]
             [status-im.utils.datetime :as time]))
 
-(defn- calculate-priority [{:keys [chats current-public-key]
-                            :contacts/keys [contacts]}
+(reg-sub :get-discoveries :discoveries)
```
Comment (yenda) [inline src/status_im/ui/screens/discover/subs.cljs:5]: can you remove this file from your commit so that there is no conflict with #2483 ?

Comment (janherich) [reply src/status_im/ui/screens/discover/subs.cljs:5]: Sure

**Pull request merged**